// boilerplate-test-db/boilerplate-test-db.js
'use strict';
const qtoolsGen = require('qtools');
const qtools = new qtoolsGen(module);
const async = require('async');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {

	qtools.validateProperties({
		subject: args || {},
		targetScope: this, //will add listed items to targetScope
		propList: [
			{
				name: 'model',
				optional: false
			},
			{
				name: 'dbName',
				optional: false
			},
			{
				name: 'callback',
				optional: true
			},
			{
				name: 'config',
				optional: false
			}
		]
	});

	if (!this.callback) {
		this.callback = () => {
			console.log(`${this.dbName} init has run`);
		}
	}

	//LOCAL VARIABLES ====================================

	//LOCAL FUNCTIONS ====================================

	const startSystem = (callback) => {
		this.model.find({
			debugDataVersion: debugDataVersion
		}, (err, result) => {
			if (!result.length) {
				this.initializeDb(callback);
			} else {
				console.log(`test ${this.dbName} is in place`);
			}
		});
	};

	//METHODS AND PROPERTIES ====================================

	this.shutdown = (message, callback) => {
		console.log(`
shutting down ${__dirname}`);
		callback('', message);
	}

	//API ENDPOINTS ====================================

	//INITIALIZATION ====================================

	const debugDataVersion = 131;

	const debugDataList = [
		{
			refId: 'condition1',
			shortName: 'condition1',
			title: 'Condition Number One',
			diagnoses: [
				{
					refId: "nursingDiagnosis 1.1",
					nursingDiagnosis: "nursingDiagnosis 1.1",
					goals: "goals 1.1",
					interventions: "interventions 1.1",
					outcomes: "outcomes 1.1",
					shortName: "shortName 1.1"
				},
				{
					refId: "nursingDiagnosis 1.2",
					nursingDiagnosis: "nursingDiagnosis 1.2",
					goals: "goals 1.2",
					interventions: "interventions 1.2",
					outcomes: "outcomes 1.2",
					shortName: "shortName 1.2"
				}

			]
		},
		{
			refId: 'condition2',
			shortName: 'condition2',
			title: 'Condition Number Two',
			diagnoses: [
				{
					refId: "nursingDiagnosis 2.1",
					nursingDiagnosis: "nursingDiagnosis 2.1",
					goals: "goals 2.1",
					interventions: "interventions 2.1",
					outcomes: "outcomes 2.1",
					shortName: "shortName 2.1"
				},
				{
					refId: "nursingDiagnosis 2.2",
					nursingDiagnosis: "nursingDiagnosis 2.2",
					goals: "goals 2.2",
					interventions: "interventions 2.2",
					outcomes: "outcomes 2.2",
					shortName: "shortName 2.2"
				}

			]
		},
		{
			refId: 'condition3',
			shortName: 'condition3',
			title: 'Condition Number Three',
			diagnoses: [
				{
					refId: "nursingDiagnosis 3.1",
					nursingDiagnosis: "nursingDiagnosis 3.1",
					goals: "goals 3.1",
					interventions: "interventions 3.1",
					outcomes: "outcomes 3.1",
					shortName: "shortName 3.1"
				},
				{
					refId: "nursingDiagnosis 3.2",
					nursingDiagnosis: "nursingDiagnosis 3.2",
					goals: "goals 3.2",
					interventions: "interventions 3.2",
					outcomes: "outcomes 3.2",
					shortName: "shortName 3.2"
				}

			]
		},
		{
			refId: 'condition4',
			shortName: 'condition4',
			title: 'Condition Number Four',
			diagnoses: [
				{
					refId: "nursingDiagnosis 4.1",
					nursingDiagnosis: "nursingDiagnosis 4.1",
					goals: "goals 4.1",
					interventions: "interventions 4.1",
					outcomes: "outcomes 4.1",
					shortName: "shortName 4.1"
				},
				{
					refId: "nursingDiagnosis 4.2",
					nursingDiagnosis: "nursingDiagnosis 4.2",
					goals: "goals 4.2",
					interventions: "interventions 4.2",
					outcomes: "outcomes 4.2",
					shortName: "shortName 4.2"
				}

			]
		},
	];

	let debugDataInx = 0;
	let userData = {};

	const addDocument = (callback) => {
		var user = new this.model(debugDataList[debugDataInx]);

		debugDataInx = debugDataInx + 1;
		user.save((err, result) => {
			if (err) {
				return console.error(err);
			}
			if (debugDataInx < debugDataList.length) {
				setTimeout(addDocument, 100, callback);
			} else {
				if (callback) {
					callback(err, result);
				}
			}
		});

	}
	this.initializeDb = (callback) => {
		if (!this.config || !this.config.database || this.config.database.allowReinit != 'true') {
			callback(`database re-init not allowed for ${this.dbName} on this server`);
			return;
		}
		
		debugDataInx = 0;
		console.log(`Clearing ${this.dbName} database`);
		this.model.remove({}, (err, result) => {
			var tmp = new this.model({
				debugDataVersion: debugDataVersion
			});
			tmp.save(() => {
				console.log('updated debugDataVersion to ' + debugDataVersion);
			});
			this.model.count((err, result) => {
				console.log(`INITIALIZING: TEST ${this.dbName.toUpperCase()}`);
				addDocument(callback);
			});
		});

	};

	startSystem(this.callback);

	return this;
};

//END OF moduleFunction() ============================================================

module.exports = moduleFunction;
//module.exports = new moduleFunction();
