// boilerplate-test-db/boilerplate-test-db.js
'use strict';
const qtoolsGen = require('qtools');
const qtools = new qtoolsGen(module);
const async = require('async');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {

	qtools.validateProperties({
		subject: args || {},
		targetScope: this, //will add listed items to targetScope
		propList: [
			{
				name: 'model',
				optional: false
			},
			{
				name: 'dbName',
				optional: false
			},
			{
				name: 'callback',
				optional: true
			},
			{
				name: 'config',
				optional: false
			}
		]
	});

	if (!this.callback) {
		this.callback = () => {
			console.log(`${this.dbName} init has run`);
		}
	}

	//LOCAL VARIABLES ====================================

	//LOCAL FUNCTIONS ====================================

	const startSystem = (callback) => {
		this.model.find({
			debugDataVersion: debugDataVersion
		}, (err, result) => {
			if (!result.length) {
				this.initializeDb(callback);
			} else {
				console.log(`test ${this.dbName} is in place`);
			}
		});
	};

	//METHODS AND PROPERTIES ====================================

	this.shutdown = (message, callback) => {
		console.log(`
shutting down ${__dirname}`);
		callback('', message);
	}

	//API ENDPOINTS ====================================

	//INITIALIZATION ====================================

	const debugDataVersion = 134;

	const debugDataList = [
		{
			refId: 'condition1',
			shortName: 'condition1',
			title: 'Condition Number One',
			diagnoses: [
				{
					refId: "nursingDiagnosis 1.1",
					nursingDiagnosis: "nursingDiagnosis 1.1",
					goals: "goals 1.1",
					interventions: "interventions 1.1",
					outcomes: "outcomes 1.1",
					shortName: "shortName 1.1",
					includeByDefault:true
				},
				{
					refId: "nursingDiagnosis 1.2",
					nursingDiagnosis: "nursingDiagnosis 1.2",
					goals: "goals 1.2",
					interventions: "interventions 1.2",
					outcomes: "outcomes 1.2",
					shortName: "shortName 1.2"
				}

			]
		},
		{
			refId: 'condition2',
			shortName: 'condition2',
			title: 'Condition Number Two',
			diagnoses: [
				{
					refId: "nursingDiagnosis 2.1",
					nursingDiagnosis: "nursingDiagnosis 2.1",
					goals: "goals 2.1",
					interventions: "interventions 2.1",
					outcomes: "outcomes 2.1",
					shortName: "shortName 2.1"
				},
				{
					refId: "nursingDiagnosis 2.2",
					nursingDiagnosis: "nursingDiagnosis 2.2",
					goals: "goals 2.2",
					interventions: "interventions 2.2",
					outcomes: "outcomes 2.2",
					shortName: "shortName 2.2"
				}

			]
		},
		{
			refId: 'condition3',
			shortName: 'condition3',
			title: 'Condition Number Three',
			diagnoses: [
				{
					refId: "nursingDiagnosis 3.1",
					nursingDiagnosis: "nursingDiagnosis 3.1",
					goals: "goals 3.1",
					interventions: "interventions 3.1",
					outcomes: "outcomes 3.1",
					shortName: "shortName 3.1"
				},
				{
					refId: "nursingDiagnosis 3.2",
					nursingDiagnosis: "nursingDiagnosis 3.2",
					goals: "goals 3.2",
					interventions: "interventions 3.2",
					outcomes: "outcomes 3.2",
					shortName: "shortName 3.2"
				}

			]
		},
		{
			refId: 'condition4',
			shortName: 'condition4',
			title: 'Condition Number Four',
			diagnoses: [
				{
					refId: "nursingDiagnosis 4.1",
					nursingDiagnosis: "nursingDiagnosis 4.1",
					goals: "goals 4.1",
					interventions: "interventions 4.1",
					outcomes: "outcomes 4.1",
					shortName: "shortName 4.1"
				},
				{
					refId: "nursingDiagnosis 4.2",
					nursingDiagnosis: "nursingDiagnosis 4.2",
					goals: "goals 4.2",
					interventions: "interventions 4.2",
					outcomes: "outcomes 4.2",
					shortName: "shortName 4.2"
				},

			]
		},
		
		
		{ 
"title" : "Abuse and Neglect: Maltreatment of Children and Adolescents",
 "shortName" : "Abuse&Neglect",
 "refId" : "230064ca-bbe5-4095-8478-9c9adb36c7fb",
 
 "diagnoses" : [ { "refId" : "6a591b89-75ca-4040-b913-b433756d8462",
 "nursingDiagnosis" : "Risk for infection related to:\n•\taltered skin integrity\n•\tinadequate vaccination status\n•\tinadequate nutrition",
 "goals" : "",
 "interventions" : "•\tContinue to assess student for further maltreatment after initial incident\n•\tProvide nursing care and comfort measures for injuries. ",
 "outcomes" : "The student will show signs of wound healing and improved skin integrity.\nThe student will report a decrease in pain/discomfort ",
 "shortName" : "ND 1 - Risk for infection",
					includeByDefault:true
},

 { "refId" : "7d7dd8b9-e116-4d5e-bc08-4e2e4cb5560f",
 "nursingDiagnosis" : "Risk for trauma related to:\n•\thistory of abuse (e.g., physical, psychological, sexual)\n\t-\timpaired parenting \n\t-\tabusive treatment\n",
 "goals" : "",
 "interventions" : "•\tReport suspected maltreatment to child protective services according to state law and school policy. \n\n•\tCoordinate interview of student by child protective services or police while student is in the school.\n\n•\tExplain to student the events that will occur after referral is made to child protective services.\n\n•\tContinue to assess student for further maltreatment after initial incident.",
 "outcomes" : "The student will report any maltreatment to a responsible adult (physical, emotional, psychosocial, or sexual).\n\nThe student will show signs of wound healing and improved skin integrity .\n\nThe student will report a decrease in pain/discomfort.\n\nThe student will receive family intervention services, which may include an immediate (temporary or permanent) protective environment to ensure safety.",
 "shortName" : "ND 2 - Risk for trauma",
},

 { "refId" : "2583af48-3709-4b3f-85f9-7497cc0f917c",
 "nursingDiagnosis" : "Acute pain related to:\n•\tphysical injury agent, recent bodily injury \n•\timpaired skin integrity \n•\tmalnutrition\n•\tbiological injury agent, infection",
 "goals" : "",
 "interventions" : "•\tContinue to assess student for further maltreatment after initial incident.\n\n•\tProvide nursing care and comfort measures for injuries.\n",
 "outcomes" : "The student will manage pain/discomfort successfully during school hours.",
 "shortName" : "ND 3 - Acute pain ",
					includeByDefault:true
},

 { "refId" : "6e982bf1-16ab-452f-b00b-0b251b8f3e2a",
 "nursingDiagnosis" : "Imbalanced nutrition less than body requirements related to:\n•\thistory of abuse (e.g., physical, psychological, sexual)\n•\tinsufficient dietary intake",
 "goals" : "",
 "interventions" : "•\tContinue to assess student for further maltreatment after initial incident.\n\n",
 "outcomes" : "The student will meet nutritional needs while in school.",
 "shortName" : "ND 4 - Imbalanced nutrition less than body requirements ",
},

 { "refId" : "27112f9b-83dd-4ace-a154-e4298a08b7c9",
 "nursingDiagnosis" : "Risk for suicide related to:\n•\thistory of abuse (e.g., physical, psychological, sexual)\n•\tpowerlessness\n•\tdepression\n•\tsocial isolation\n•\tdisruptive family life\n•\tinsufficient social support ",
 "goals" : "",
 "interventions" : "•\tContinue to assess student for further maltreatment after initial incident. \n\n•\tDevelop and maintain a trusting therapeutic relationship with student, offering emotional support as needed.\n\n•\tProvide a safe haven for student in the health suite when the stresses of maltreatment cause anxiety and dysfunctional behavior. \n\n•\tSupport student and family involvement in mental health treatment, counseling, and support groups.\n\n",
 "outcomes" : "The student will verbalize feelings regarding family relationships, home environment, and maltreatment issues.\n\nThe student will maintain therapeutic trusting relationships with school professionals.\n\nThe student will participate in school-based counseling when offered as a service.\n\n",
 "shortName" : "ND 5 - Risk for suicide ",
},

 { "refId" : "8bde335e-9891-449f-9ae8-9c90cdfe4d45",
 "nursingDiagnosis" : "ND 6 Impaired social interaction related to:\n•\thistory of abuse (e.g., physical, psychological, sexual)\n•\tdisturbance in self-concept\n\t-\tenvironmental barrier\n\t-\tinadequate support system",
 "goals" : "",
 "interventions" : "•\tContinue to assess student for further maltreatment after initial incident. \n\n•\tDevelop and maintain a trusting therapeutic relationship with student, offering emotional support as needed.\n\n•\tRefer child for supportive therapy.\n\n•\tProvide a safe haven for student in the health suite when the stresses of maltreatment cause anxiety and dysfunctional behavior.\n\n•\tEncourage and help student develop peer relationships: conversation starters, role play, brainstorming, joining school-sponsored groups.\n\n•\tRefer student to support group for children dealing with abuse/neglect issues.",
 "outcomes" : "The student will develop and maintain at least two friendships with peers.\n\nThe student will verbalize feelings regarding family relationships, home environment, and maltreatment issues.\n\nThe student participate in school and recreational activities with peers.\n\nThe student will achieve academic success appropriate to cognitive abilities.\n\nThe student will maintain therapeutic trusting relationships with school professionals.\n\nThe student will participate in school-based counseling when offered as a service.\n",
 "shortName" : "ND 6 - Impaired social interaction",
},

 { "refId" : "c2250195-d3cc-45c8-b50d-513f064364ab",
 "nursingDiagnosis" : "ND 7 Ineffective protection related to:\n•\thistory of abuse, (e.g., physical, psychological, sexual)\n•\timpaired parenting\n•\tinadequate nutrition\n",
 "goals" : "",
 "interventions" : "•\tExplain to student the events that will occur after referral is made to child protective services.\n\n•\tRemain with child during interview and provide emotional support.\n\n•\tContinue to assess student for further maltreatment after initial incident.\n\n•\tProvide a safe haven for student in the health suite when the stresses of maltreatment cause anxiety and dysfunctional behavior.\n\n•\tSupport student and family involvement in mental health treatment, counseling, and support groups. ",
 "outcomes" : "The student will develop and use emergency plan when in presence of danger.\n\nThe student will verbalize feelings regarding family relationships, home environment, and maltreatment issue.\n\nThe student will maintain therapeutic trusting relationships with school professionals.\n\nThe student will receive family intervention services, which may include an immediate (temporary or permanent) protective environment to ensure safety.",
 "shortName" : "ND 7 - Ineffective protection ",
},

 { "refId" : "90be61b0-039e-4367-ac0e-24370e0e9ddd",
 "nursingDiagnosis" : "ND 8 Anxiety related to:\n•\thistory of abuse (e.g., physical, psychological, sexual)\n•\tunmet needs\n•\tself-concept disturbance\n•\tinadequate support systems\n•\tthreat of injury or death",
 "goals" : "",
 "interventions" : "•\tRemain with child during interview and provide emotional support.\n\n•\tContinue to assess student for further maltreatment after initial incident.\n\n•\tDevelop and maintain a trusting therapeutic relationship with student, offering emotional support as needed.\n\n•\tProvide a safe haven for student in the health suite when the stresses of maltreatment cause anxiety and dysfunctional behavior.\n\n•\tEncourage and help student develop peer relationships: conversation starters, role play, brainstorming, joining school-sponsored groups. \n\n•\tSupport student and family involvement in mental health treatment, counseling, and support groups. \n\n•\tProvide encouragement for hard work and achievement.\n\n•\tRefer student for child student team evaluation, if appropriate.\n\n•\tMonitor academic progress and need for further support.\n",
 "outcomes" : "The student will develop and use emergency plan when in presence of danger.\n\nThe student will participate in school-based counseling when offered as a service.\n\nThe student will receive family intervention services, which may include an immediate (temporary or permanent) protective environment to ensure safety.\n\nThe student will improve grades within one marking.",
 "shortName" : "ND 8 - Anxiety",
},

 { "refId" : "431bbd04-6d80-4d21-929c-92e6322255f5",
 "nursingDiagnosis" : "ND 9 Impaired parenting related to:\n•\tpresence of stressor (e.g., financial problems, legal issues, family crisis)\n\tstress of dysfunctional family interactions\n•\trole strain\n•\tinsufficient cognitive readiness for parenting\n•\tcompromised home environment\n•\thistory of abuse (e.g., physical, psychological, sexual)\n•\tinsufficient parenting skills\n•\tunrealistic expectations for child, self, or others\n•\tpreference for physical punishment\n•\thistory of substance abuse\n•\thistory of mental illness",
 "goals" : "",
 "interventions" : "•\tContinue to assess student for further maltreatment after initial incident.\n\n•\tSupport student and family involvement in mental health treatment, counseling, and support groups.",
 "outcomes" : "The student will develop and use emergency plan when in presence of danger.\n\nThe student will verbalize feelings regarding family relationships, home environment, and maltreatment issues.\n\nThe student will maintain therapeutic trusting relationships with school professionals.\n\nThe student will participate in school-based counseling when offered as a service.\n\nThe student will receive family intervention services, which may include an immediate (temporary or permanent) protective environment to ensure safety.\n",
 "shortName" : "ND 9 - Impaired parenting",
},

 { "refId" : "97d12aa9-7df2-4a7f-a771-d7a10ab0bf03",
 "nursingDiagnosis" : "Ineffective role performance related to:\n•\tdomestic violence and conflict (history of abuse [e.g., physical, psychological, sexual])\n•\tlow self-esteem\n•\tmental health issues or diagnoses\n•\tpain\t\n•\tinsufficient resources\n\t-\tlack of knowledge regarding role skills \n•\tinsufficient support system\n\t-\tinsufficient role preparation\n\t-\tunrealistic role expectations\n",
 "goals" : "",
 "interventions" : "•\tContinue to assess student for further maltreatment after initial incident.\n\n•\tDevelop and maintain a trusting therapeutic relationship with student, offering emotional support as needed.\n\n•\tRefer child for supportive therapy. \n\n•\tEncourage and help student develop peer relationships: conversation starters, role play, brainstorming, joining school-sponsored groups.\n\n•\tSupport student and family involvement in mental health treatment, counseling, and support groups.\n\n•\tProvide encouragement for hard work and achievement.\n\n•\tRefer student for child student team evaluation, if appropriate.\n\n•\tMonitor academic progress and need for further support.\n\n•\tRefer student for academic support and/or peer tutoring and support when grades are affected by home situation. ",
 "outcomes" : "The student will verbalize feelings regarding family relationships, home environment, and maltreatment issues.\n\nThe student will participate in school and recreational activities with peers.\n\nThe student will achieve academic success appropriate to cognitive abilities.\n\nThe student will maintain therapeutic trusting relationships with school professionals.\n\nThe student will participate in school-based counseling when offered as a service.\n\nThe student will improve grades within one marking.",
 "shortName" : "ND 10 - Ineffective role performance",
},

 { "refId" : "f232a7a4-60f8-497e-8721-2fead76e06e0",
 "nursingDiagnosis" : "Readiness for enhanced knowledge related to:\n•\texpressed need to enhance learning",
 "goals" : "",
 "interventions" : "•\tTeach student health maintenance skills: nutrition, care of wounds, sleep-promoting behaviors, stress management, and personal body protection.",
 "outcomes" : "The student will participate in school and recreational activities with peers.\n\nThe student will maintain therapeutic trusting relationships with school professionals.\n\nThe student will participate in school-based counseling when offered as a service.\n\nThe student will improve grades within one marking.\n",
 "shortName" : "ND 11 - Readiness for enhanced knowledge",
}
]
}
	];


var addConditions=()=>{
let newItem;
for (var i=100; i<170; i++){
	newItem={
			refId: qtools.newGuid(),
			shortName: 'condition'+i,
			title: 'Condition Number #'+i,
			diagnoses: [
				{
					refId: qtools.newGuid(),
					nursingDiagnosis: "nursingDiagnosis "+i+".1",
					goals: "goals "+i+".1",
					interventions: "interventions "+i+".1",
					outcomes: "outcomes "+i+".1",
					shortName: "shortName "+i+".1"
				},
				{
					refId: qtools.newGuid(),
					nursingDiagnosis: "nursingDiagnosis "+i+".2",
					goals: "goals "+i+".2",
					interventions: "interventions "+i+".2",
					outcomes: "outcomes "+i+".2",
					shortName: "shortName "+i+".2"
				},

			]
		}
	debugDataList.push(newItem);
}
}
addConditions();

	let debugDataInx = 0;
	let userData = {};

	const addDocument = (callback) => {
		var user = new this.model(debugDataList[debugDataInx]);

		debugDataInx = debugDataInx + 1;
		user.save((err, result) => {
			if (err) {
				return console.error(err);
			}
			if (debugDataInx < debugDataList.length) {
				setTimeout(addDocument, 100, callback);
			} else {
				if (callback) {
					callback(err, result);
				}
			}
		});

	}
	this.initializeDb = (callback) => {
		if (!this.config || !this.config.database || this.config.database.allowReinit != 'true') {
			callback(`database re-init not allowed for ${this.dbName} on this server`);
			return;
		}
		
		debugDataInx = 0;
		console.log(`Clearing ${this.dbName} database`);
		this.model.remove({}, (err, result) => {
			var tmp = new this.model({
				debugDataVersion: debugDataVersion
			});
			tmp.save(() => {
				console.log('updated debugDataVersion to ' + debugDataVersion);
			});
			this.model.count((err, result) => {
				console.log(`INITIALIZING: TEST ${this.dbName.toUpperCase()}`);
				addDocument(callback);
			});
		});

	};

	startSystem(this.callback);

	return this;
};

//END OF moduleFunction() ============================================================

module.exports = moduleFunction;
//module.exports = new moduleFunction();
