// boilerplate/boilerplate.js
'use strict';
const qtoolsGen = require('qtools');
const qtools = new qtoolsGen(module);
//const crypt=require('crypto');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {

	qtools.validateProperties({
		subject: args || {},
		targetScope: this, //will add listed items to targetScope
		propList: [
			{
				name: 'config',
				optional: false
			},
			{
				name: 'router',
				optional: false
			},
			{
				name: 'mongoose',
				optional: false
			},
			{
				name: 'permissionMaster',
				optional: false
			},
			{
				name: 'initCallback',
				optional: false
			}
		]
	});

	//LOCAL VARIABLES ====================================
	const mongoose = this.mongoose;

	//LOCAL FUNCTIONS ====================================

	//METHODS AND PROPERTIES ====================================

	this.shutdown = (message, callback) => {
		console.log(`shutting down ${__dirname}`);
		callback('', message);
	}

	this.get = (criteria, callback) => {
		Boilerplate.find(criteria, callback);

	}

	//API ENDPOINTS ====================================

	let route = new RegExp('boilerplate$');
	this.permissionMaster.addRoute(route, 'all');
	this.router.get(route, (req, res, next) => {

		this.get({}, (err, users) => {
			if (err) {
					next({
						code: '500',
						message: 'Database access problem (boilerplate.get())'
					});
			} else {
				res.json({token:this.permissionMaster.updateToken(req), data:users});
			}
		});

	});

	this.router.post(/user/, (req, res, next) => {

		res.set({
			'content-type': 'application/json;charset=ISO-8859-1',
			messageid: qtools.newGuid(),
			messagetype: 'RESPONSE',
			responsesource: 'utilityServer',
			connection: 'Close'
		});

		res.json({
			status: `CONNECTIVITY ESTABLISHED FROM ${this.config.system.name}/${this.config.user}${req.path} GET`,
			headers: req.headers,
			body: req.body,
			query: req.query
		});
	});

	//INITIALIZATION ====================================

	const boilerplateSchema = mongoose.Schema({
		shortName: String,
		title: String,
		debugDataVersion: String
	});

	boilerplateSchema.methods.speak = function() {
		const greeting = this.name
			? "Meow name is " + this.name
			: "I don't have a name";
		console.log(greeting);
	}

	const Boilerplate = mongoose.model('Boilerplate', boilerplateSchema);

	const debugDataList = [
		{
			shortName: 'condition1',
			title: 'Condition Number One'
		},
		{
			shortName: 'condition2',
			title: 'Condition Number Two'
		},
		{
			shortName: 'condition3',
			title: 'Condition Number Three'
		},
		{
			shortName: 'condition4',
			title: 'Condition Number Four'
		},
	];
	const debugDataVersion = 120;

	let debugDataInx = 0;
	let userData = {};

	const addBoilerplate = () => {

		var user = new Boilerplate(debugDataList[debugDataInx]);

		debugDataInx = debugDataInx + 1;

		user.save((err, result) => {
			if (err) {
				return console.error(err);
			}
			if (debugDataInx < debugDataList.length) {
				setTimeout(addBoilerplate, 100);
			}
		});

	}
	
	Boilerplate.find({
		debugDataVersion: debugDataVersion
	}, (err, result) => {
		if (!result.length) {
			console.log('Clearing boilerplate database');
			Boilerplate.remove({}, (err, result) => {
				var tmp = new Boilerplate({
					debugDataVersion: debugDataVersion
				});
				tmp.save(() => {
					console.log('updated debugDataVersion to ' + debugDataVersion);
				});
				Boilerplate.count((err, result) => {
					console.log('INITIALIZING: TEST BOLIERPLATE');
					addBoilerplate();
				});
			});
		} else {
			console.log('test boilerplate is in place');
		}
	});

	this.initCallback();
	return this;
};

//END OF moduleFunction() ============================================================

module.exports = moduleFunction;
//module.exports = new moduleFunction();
