// mt-collector/mt-collector.js
'use strict';
const qtoolsGen = require('qtools');
const qtools = new qtoolsGen(module);
const async = require('async');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {

	qtools.validateProperties({
		subject: args || {},
		targetScope: this, //will add listed items to targetScope
		propList: [
			{
				name: 'config',
				optional: false
			},
			{
				name: 'database',
				optional: false
			},
			{
				name: 'apiManager',
				optional: false
			},
			{
				name: 'initCallback',
				optional: true
			}
		]
	});

	//LOCAL VARIABLES ====================================

	let workerList = {};
	
	//LOCAL FUNCTIONS ====================================

	const startSystem = () => {

		if (typeof (COMPONENT) == 'undefined') {
			this.initCallback && this.initCallback();
			return;
		}

		const startList = [];

		startList.push((done) => {
			const workerName = 'COMPONENT_NAME'
			new COMPONENT({
				initCallback: function() {
					workerList[workerName] = this; done();
				}
			});
		});

		async.series(startList, () => {
			this.initCallback && this.initCallback();
		});
	};

	//METHODS AND PROPERTIES ====================================

	//API ENDPOINTS ====================================

	//INITIALIZATION ====================================
	
	setTimeout(()=>{
		console.log("mail-tools.collector says, not sending test email");
		return;
		this.apiManager.getApi('mailTool.messages.addMessage')({
			clientRefId:'collector.test',
			emailAddress:'tqwhite@erdc.k12.mn.us',
			schedule:'now', 
			templateName:'confirmEmail', 
			dictionary:[
			{
				pattern: 'first',
				replacement: 'TQ'
			},


			{
				pattern: 'last',
				replacement: 'White II'
			},


			{
				pattern: 'userRefId',
				replacement: "tqwhite"
			},


			{
				pattern: 'time',
				replacement: new Date()
			}
			]
		
		});
	}, 2000);

	startSystem();


	//SHUTDOWN FUNCTIONS ====================================	

	this.shutdown = (message, callback) => {
		console.log(`\nshutting down ${__dirname}`);
		callback('', message);
	}
	
	return this;
};

//END OF moduleFunction() ============================================================

module.exports = moduleFunction;
//module.exports = new moduleFunction();
