// plans-test-db/plans-test-db.js
'use strict';
const qtoolsGen = require('qtools');
const qtools = new qtoolsGen(module);
const async = require('async');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {

	qtools.validateProperties({
		subject: args || {},
		targetScope: this, //will add listed items to targetScope
		propList: [
			{
				name: 'oneToManyAccessModel',
				optional: false
			},
			{
				name: 'oneToManyAccessModelName',
				optional: true
			},
			{
				name: 'model',
				optional: false
			},
			{
				name: 'dbName',
				optional: false
			},
			{
				name: 'callback',
				optional: true
			},
			{
				name: 'config',
				optional: false
			}
		]
	});

	if (!this.callback) {
		this.callback = () => {
			console.log(`${this.dbName} init has run`);
		}
	}

	//LOCAL VARIABLES ====================================

	//LOCAL FUNCTIONS ====================================

	const startSystem = (callback) => {
		this.model.find({
			debugDataVersion: debugDataVersion
		}, (err, result) => {
			if (!result.length) {
				this.initializeDb(callback);
			} else {
				console.log(`test ${this.dbName} is in place`);
			}
		});
	};

	//METHODS AND PROPERTIES ====================================

	this.shutdown = (message, callback) => {
		console.log(`
shutting down ${__dirname}`);
		callback('', message);
	}

	//API ENDPOINTS ====================================

	//INITIALIZATION ====================================

	const debugDataVersion = 148;

	const sourceList = [
		{
			STUDENTREFID:'Addison_White',
  refId: "ad",
  conditions: [
    {
      refId: "fbba12b4-e30e-4eb9-b6c1-1ebbc999d75d",
      sourceConditionRefId: "25d30d28-c095-40db-9bf0-e769518e2441",
      title: "Arthritis, Juvenile",
      diagnoses: [
        {
          refId: "b2e7788a-fe7c-4312-aa46-a32a91dc1f2d",
          sourceDiagnosisRefId: "a06ff0c5-e82a-415b-b85b-6877a3af5cc4",
          nursingDiagnosis: "Chronic pain syndrome related to:\n•\timpaired physical mobility\n•\tsocial isolation\n•\temotional stress (e.g., anxiety, fear)\n•\tfatigue",
          interventions: "The school nurse will monitor for signs/symptoms of pain (e.g., standard assessment tools, student, parent reports of pain and activity level).\n\nThe school nurse will provide comfort measures (e.g., medications as directed, rest, range of motion, heat) as directed.\n\nThe school nurse will monitor for medication side effects and adverse reactions.\n\nThe school nurse will monitor assistive orthotics and/or refer to physical or occupational therapy for comfort (e.g., splints for upper/lower extremities).\n\nThe school nurse will monitor assistive devices for function and condition and report to appropriate persons for any needed adjustments or repair.\n \nThe school nurse will (teach, monitor, reinforce) safety plan measures to be observed in the school environment related to __________ (e.g., wheelchair, braces, companion dog).\n\nThe school nurse will provide accommodations or modifications for educational instruction and expectations.\n\nThe school nurse will problem-solve ways to meet nutritional and fluid needs.\n\nThe school nurse will facilitate plan for service dog use, care, and interactions with student and others in the school community. ",
          outcomes: "The student will participate to fullest extent with minimal or limited impact as a result of JIA _____ % of the time.\n\nThe student will have all equipment and orthotics available and in working order ________% of the time.\n\nThe student will conveniently access comfort measures with limited interruption in instruction and peer social time ______% of the time.\n\nThe student will identify early signs and symptoms of pain and promptly seek assistance from school nurses or other designated school personnel.\n\nThe student will safely navigate the school environment with assistive equipment.\n\nThe student will access pain relief measures and toileting during natural transition time _____% of the time.\n",
          shortName: "ND 2 - Chronic pain syndrome"
        },
        {
          refId: "7ad1d27f-f7a6-49bb-8169-16c8405c0bc5",
          sourceDiagnosisRefId: "e6c32c87-4f49-4aac-aaba-2ba43c70634f",
          nursingDiagnosis: "Risk of infection related to:\n•\tchronic illness\n•\timmunosuppression\n•\tsuppressed inflammatory response (i.e., medication therapies)",
          interventions: "The school nurse will monitor for medication side effects and adverse reactions.\n",
          outcomes: "The student will safely navigate the school environment with assistive equipment.",
          shortName: "ND 3 - Risk of infection"
        }
      ]
    }
  ]

		},
		{
			STUDENTREFID:'Emily_White',
			refId:'em',

  refId: "em",
  conditions: [
    {
      refId: "51af96d5-5fbd-4dfe-8fe5-424dcd13e365",
      sourceConditionRefId: "6457906c-4c4a-43fb-b127-4590664d36dd",
      title: "Encopresis",
      diagnoses: [
        {
          refId: "733c6596-3cdf-4988-9e42-29df69a95621",
          sourceDiagnosisRefId: "7c6367f9-e292-488c-bd16-838cd8b9e3e9",
          nursingDiagnosis: "Toileting self-care deficit related to:\n•\tenvironmental barrier\n•\tdecrease in or lack of motivation\n•\tanxiety\n•\tpain or discomfort\n•\taltered cognitive functioning or perceptual impairment ",
          interventions: "The school nurse will institute a toileting schedule at school. Arrange for student to come to nurse’s office after lunch every day for a toilet sit and, if necessary, at the beginning of each school day if student is unable to toilet sit at home before school. \n\nThe school nurse will ensure that student performs daily toilet sits using proper positioning as scheduled. Teach student to bear down using the technique of blowing on a balloon while sitting on toilet and using foot support. \n\nThe school nurse will arrange for permission for student to use the bathroom any time student requests.\n\nThe school nurse will encourage student to be independent in performing toileting and hygiene care. ",
          outcomes: "The student will toilet independently. \n\nThe student will stay soil-free throughout the school day. \n\nThe student will position self on toilet to facilitate bowel movement with each toileting. \n\nThe student will wash hands after toileting is complete. ",
          shortName: "ND 4 - Toileting self-care deficit"
        },
        {
          refId: "479447ee-97a1-4f07-aac1-f52d449e72a0",
          sourceDiagnosisRefId: "9b885282-adc7-4473-9643-949e3bc48e89",
          nursingDiagnosis: "Situational low self-esteem related to:\n•\tdevelopmental transitions or changes\n•\talteration in body image\n•\tfunctional impairment (soiling)\n•\tpattern of failures or history of rejections",
          interventions: "The school nurse will convey confidence in student’s ability to handle situation.\n\nThe school nurse will encourage participation in school activities.",
          outcomes: "The student will stay soil-free throughout the school day.\n\nThe student will participate in developmentally appropriate peer group social activities.",
          shortName: "ND 5 - Situational low self-esteem"
        }
      ]
    },
    {
      refId: "c4832488-ad3b-47ea-a224-f7b2807deef2",
      sourceConditionRefId: "ba4f2592-ca97-4ed7-81fe-be66e664e3c4",
      title: "Eating Disorders",
      diagnoses: [
        {
          refId: "c62afdb2-e6cf-495d-9c8f-0ab3754916a6",
          sourceDiagnosisRefId: "3a49b6c4-91f4-487c-a9f6-88bad3442003",
          nursingDiagnosis: "Disturbed body image related to:\n•\talteration in self perception\n•\tdevelopmental transition\n•\timpaired psychological functioning",
          interventions: "The school nurse will establish therapeutic relationship by allowing access to the health office as needed, providing active listening to help student express anxieties regarding food and feelings of anxiety, provide evidence-based reflective feedback to the student in a safe and supportive environment, and avoid affirming distorted perceptions. Identify and encourage positive qualities, traits, and abilities of the student to help build self-esteem. \n\nThe school nurse will report disturbed and distorted statements as well as health-enhancing statements to therapist/parent per agreement and treatment plan.\n\nThe school nurse will monitor for effectiveness, side effects, and adherence to medication.\n\nThe school nurse will enlist counseling staff or trusted cafeteria employee to monitor food intake as appropriate for staffing. \n\nThe school nurse will educate the school staff, including food service team and coaches, on disordered eating, healthy eating behaviors, the emotions surrounding eating, the role of stress in the classroom in disordered eating, and the need to monitor and adjust activity levels for recovering students.\n\nThe school nurse will include consideration of disordered eating when appropriate in Positive Behavior Intervention and Support Team and Response to Intervention meetings.\n\nThe school nurse will provide ongoing communication to parents and the treatment team regarding thought processes and behavior, with appropriate release of information as directed in the treatment plan. \n\nThe school nurse will facilitate education about eating disorders in the classroom setting. ",
          outcomes: "The student will consent to and participate in the development of the school treatment plan within __ days of disclosure to the school.\n\nThe student will sign contract assuring that he or she will disclose thoughts of self-harm or suicide to an identified individual effective upon signage. \n\nThe student will engage in prescribed levels of academic, physical, and social activities ___% of the time after signing the treatment plan. \n\nThe student will allow discrete monitoring of food intake for reports to physician/therapist ___% of the time upon signing the treatment plan. \n\nThe student will participate in the therapeutic relationship with school staff by identifying stressors in the school setting and developing accommodations to increase effective coping and adherence to the treatment plan during weekly evaluation meetings with the school nurse or other identified staff. \n\nThe student will adhere to prescribed medication regimen ___% of the time while at school. ",
          shortName: "ND 4 - Disturbed body image"
        },
        {
          refId: "f07dde49-b4ce-4575-a098-0ab5f63c9327",
          sourceDiagnosisRefId: "8b52775d-49bf-44bb-892a-7ad6bb391e51",
          nursingDiagnosis: "Deficient knowledge related to:\n•\talteration in cognitive functioning\n•\tinsufficient information\n•\tinsufficient knowledge of resources\n•\tmisinformation presented by others",
          interventions: "The school nurse will treat food consumption as a medication order, providing a safe environment for eating while discretely monitoring food intake and supporting healthy food and hydration choices. Monitor purging activities.\n\nThe school nurse will establish therapeutic relationship by allowing access to the health office as needed, providing active listening to help student express anxieties regarding food and feelings of anxiety, provide evidence-based reflective feedback to the student in a safe and supportive environment, and avoid affirming distorted perceptions. Identify and encourage positive qualities, traits, and abilities of the student to help build self-esteem.  \n\nThe school nurse will supervise release for participation in physical education, sports, and other school activities, monitoring for signs and symptoms of intolerance and providing appropriate rest periods. Provide diversional activities that are calming to the student.\n\nThe school nurse will report disturbed and distorted statements as well as health-enhancing statements to therapist/parent per agreement and treatment plan.\n\nThe school nurse will enlist counseling staff or trusted cafeteria employee to monitor food intake as appropriate for staffing. \n\nThe school nurse will educate the school staff, including food service team and coaches, on disordered eating, healthy eating behaviors, the emotions surrounding eating, the role of stress in the classroom in disordered eating, and the need to monitor and adjust activity levels for recovering students. \n\nThe school nurse will include consideration of disordered eating when appropriate in Positive Behavior Intervention and Support Team and Response to Intervention meetings. \n\nThe school nurse will provide ongoing communication to parents and the treatment team regarding thought processes and behavior, with appropriate release of information as directed in the treatment plan. \n \nThe school nurse will facilitate education about eating disorders in the classroom setting. ",
          outcomes: "The student will consent to and participate in the development of the school treatment plan within __ days of disclosure to the school. \n\nThe student will allow discrete monitoring of food intake for reports to physician/therapist ___% of the time upon signing the treatment plan. \n\nThe student will adhere to prescribed medication regimen ___% of the time while at school. ",
          shortName: "ND 10 - Deficient knowledge"
        }
      ]
    }
  ]

		}
	];

	const debugDataList = [];
	const accessList = {};
	const debugAccessList = [];

	for (var i = 0, len = sourceList.length; i < len; i++) {
		var element = sourceList[i];
		var ownerRefId = element.STUDENTREFID;
		element.refId = element.refId;

		if (typeof (accessList[ownerRefId]) == 'undefined') {
			accessList[ownerRefId] = {
				refId: qtools.newGuid(),
				studentRefId: ownerRefId,
				planRefIdList: []
			}
		}

		accessList[ownerRefId].planRefIdList.push(element.refId);
		delete element.user;
		debugDataList.push(element);
	}

	for (var i in accessList) {
		var element = accessList[i];
		debugAccessList.push(element);
	}

	const addDocument = (callback) => {
		const localCallback = function() {
			addAccess(callback);
		}
		addPrimary(localCallback);

	}

	const addPrimary = (callback) => {

		const taskList = [];

		for (var i = 0, len = debugDataList.length; i < len; i++) {
			var element = debugDataList[i];

			taskList.push(function(item) {
				return (next) => {
					new this.model(item).save(next);
					return;
				}
			}.bind(this)(element));
		}
		async.series(taskList, callback);
		return;
	}
	
	const addAccess = (callback) => {

		const taskList = [];

		for (var i = 0, len = debugAccessList.length; i < len; i++) {
			var element = debugAccessList[i];

			taskList.push(function(item) {
				return (next) => {
					new this.oneToManyAccessModel(item).save(next);
					return;
				}
			}.bind(this)(element));
		}
		async.series(taskList, callback);
		return;
	}

	this.initializeDb = (callback) => {
		if (!this.config || !this.config.database || this.config.database.allowReinit != 'true') {
			callback(`database re-init not allowed for ${this.dbName} on this server`);
			return;
		}

		console.log(`Clearing ${this.dbName} database`);
		this.model.remove({}, (err, result) => {

			this.oneToManyAccessModel.remove({}, (err, result) => {
				var tmp = new this.model({
					debugDataVersion: debugDataVersion
				});
				tmp.save(() => {
					console.log('updated debugDataVersion to ' + debugDataVersion);

					this.model.count((err, result) => {
						console.log(`INITIALIZING: TEST ${this.dbName.toUpperCase()} ${this.oneToManyAccessModelName && this.oneToManyAccessModelName.toUpperCase()}`);
						addDocument(callback);
					});
				});
			});
		});

	};

	startSystem(this.callback);

	return this;
};

//END OF moduleFunction() ============================================================

module.exports = moduleFunction;
//module.exports = new moduleFunction();
