// user-test-db/user-test-db.js
'use strict';
const qtoolsGen = require('qtools');
const qtools = new qtoolsGen(module);
const async = require('async');

//START OF moduleFunction() ============================================================

var moduleFunction = function(args) {

	qtools.validateProperties({
		subject: args || {},
		targetScope: this, //will add listed items to targetScope
		propList: [
			{
				name: 'model',
				optional: false
			},
			{
				name: 'dbName',
				optional: false
			},
			{
				name: 'callback',
				optional: true
			},
			{
				name: 'permissionMaster',
				optional: true
			},
			{
				name: 'config',
				optional: true
			}
		]
	});
	
	if (!this.callback){
		this.callback=()=>{
			console.log(`${this.dbName} init has run`);
		}
	}

	//LOCAL VARIABLES ====================================

	//LOCAL FUNCTIONS ====================================

	const startSystem = (callback) => {
		this.model.find({
			debugDataVersion: debugDataVersion
		}, (err, result) => {
			if (!result.length) {
			this.initializeDb(callback);
			} else {
				console.log(`test ${this.dbName} is in place`);
			}
		});
	};

	//METHODS AND PROPERTIES ====================================

	this.shutdown = (message, callback) => {
		console.log(`
shutting down ${__dirname}`);
		callback('', message);
	}

	//API ENDPOINTS ====================================

	//INITIALIZATION ====================================
	
	const debugDataVersion = 136;

	const debugDataList = [
		{
			refId: 'nurse2_refid',
			first: 'nurse2',
			last: 'nurse2_last',
			username: 'nurse2',
			pwhash: qtools.passwordHash('test'),
			role: this.permissionMaster.getRole('nurse').name
		},
		{
			refId: 'nurse_refid',
			first: 'nurse',
			last: 'nurse_last',
			username: 'nurse',
			pwhash: qtools.passwordHash('test'),
			role: this.permissionMaster.getRole('nurse').name
		},
		{
			refId: 'editor_refid',
			first: 'editor',
			last: 'editor_last',
			username: 'editor',
			pwhash: qtools.passwordHash('test'),
			role: this.permissionMaster.getRole('editor').name
		},
		{
			refId: 'admin_refid',
			first: 'admin',
			last: 'admin_last',
			username: 'admin',
			pwhash: qtools.passwordHash('test'),
			role: this.permissionMaster.getRole('admin').name
		},
	];

	let debugDataInx = 0;
	let userData = {};

	const addDocument = (callback) => {
		var user = new this.model(debugDataList[debugDataInx]);
		debugDataInx = debugDataInx + 1;
		user.save((err, result) => {
			if (err) {
				return console.error(err);
			}
			if (debugDataInx < debugDataList.length) {
				setTimeout(addDocument, 100, callback);
			}
			else{
				if (callback){
					callback(err, result);
				}
			}
		});

	}
	this.initializeDb=(callback)=>{
		if (!this.config || !this.config.database || this.config.database.allowReinit != 'true') {
			callback(`database re-init not allowed for ${this.dbName} on this server`);
			return;
		}
		
		debugDataInx=0;
			console.log(`Clearing ${this.dbName} database`);
			this.model.remove({}, (err, result) => {
				var tmp = new this.model({
					debugDataVersion: debugDataVersion
				});
				tmp.save(() => {
					console.log('updated debugDataVersion to ' + debugDataVersion);
				});
				this.model.count((err, result) => {
					console.log(`INITIALIZING: TEST ${this.dbName.toUpperCase()}`);
					addDocument(callback);
				});
			});
		
		};

	startSystem(this.callback);
	
	return this;
};

//END OF moduleFunction() ============================================================

module.exports = moduleFunction;
//module.exports = new moduleFunction();
